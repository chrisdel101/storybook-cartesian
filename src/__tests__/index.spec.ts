import cartesian from '../index'

describe('cartesian', () => {
  it('all valid', () => {
    const s: any[] = []
    const stories = {
      add: (title: string, story: any) => s.push({ title, ...story() })
    }
    cartesian(stories).add(
      () => ({ oneProp: [true, false], twoProp: ['', 'test'] }),
      props => `title: [${props.oneProp}], [${props.twoProp}]`,
      props => ({
        story: {
          props
        }
      })
    )
    expect(s).toMatchSnapshot()
  })
  it('some invalid', () => {
    const s: any[] = []
    const stories = {
      add: (title: string, story: any) => s.push({ title, ...story() })
    }
    cartesian(stories).add(
      () => ({
        oneProp: [true, false],
        twoProp: ['', 'foobar', 'foobaz', 'test']
      }),
      props => `title: [${props.oneProp}], [${props.twoProp}]`,
      props => ({
        story: {
          props
        }
      }),
      props => !(props.oneProp === false && props.twoProp === '')
    )
    expect(s).toMatchSnapshot()
  })
})
