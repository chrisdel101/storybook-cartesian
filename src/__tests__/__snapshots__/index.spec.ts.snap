// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cartesian choices empty prop 1`] = `Array []`;

exports[`cartesian choices multi prop 1`] = `
Array [
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": "",
      },
    },
    "title": "{\\"oneProp\\":true,\\"twoProp\\":\\"\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": "test",
      },
    },
    "title": "{\\"oneProp\\":true,\\"twoProp\\":\\"test\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": "",
      },
    },
    "title": "{\\"oneProp\\":false,\\"twoProp\\":\\"\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": "test",
      },
    },
    "title": "{\\"oneProp\\":false,\\"twoProp\\":\\"test\\"}",
  },
]
`;

exports[`cartesian choices nested prop 1`] = `
Array [
  Object {
    "story": Object {
      "props": Object {
        "oneProp": Object {
          "accept": true,
          "story": "yes",
        },
        "twoProp": "",
      },
    },
    "title": "{\\"twoProp\\":\\"\\",\\"oneProp\\":{\\"story\\":\\"yes\\",\\"accept\\":true}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": Object {
          "accept": true,
          "story": "yes",
        },
        "twoProp": "test",
      },
    },
    "title": "{\\"twoProp\\":\\"test\\",\\"oneProp\\":{\\"story\\":\\"yes\\",\\"accept\\":true}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": Object {
          "accept": false,
          "story": "yes",
        },
        "twoProp": "",
      },
    },
    "title": "{\\"twoProp\\":\\"\\",\\"oneProp\\":{\\"story\\":\\"yes\\",\\"accept\\":false}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": Object {
          "accept": false,
          "story": "yes",
        },
        "twoProp": "test",
      },
    },
    "title": "{\\"twoProp\\":\\"test\\",\\"oneProp\\":{\\"story\\":\\"yes\\",\\"accept\\":false}}",
  },
]
`;

exports[`cartesian choices single prop 1`] = `
Array [
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
      },
    },
    "title": "{\\"oneProp\\":true}",
  },
]
`;

exports[`cartesian choices some invalid 1`] = `
Array [
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": "",
      },
    },
    "title": "{\\"oneProp\\":true,\\"twoProp\\":\\"\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": "foobar",
      },
    },
    "title": "{\\"oneProp\\":true,\\"twoProp\\":\\"foobar\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": "foobaz",
      },
    },
    "title": "{\\"oneProp\\":true,\\"twoProp\\":\\"foobaz\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": "test",
      },
    },
    "title": "{\\"oneProp\\":true,\\"twoProp\\":\\"test\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": "",
      },
    },
    "title": "{\\"oneProp\\":false,\\"twoProp\\":\\"\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": "foobar",
      },
    },
    "title": "{\\"oneProp\\":false,\\"twoProp\\":\\"foobar\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": "foobaz",
      },
    },
    "title": "{\\"oneProp\\":false,\\"twoProp\\":\\"foobaz\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": "test",
      },
    },
    "title": "{\\"oneProp\\":false,\\"twoProp\\":\\"test\\"}",
  },
]
`;

exports[`cartesian legends render title 1`] = `
Array [
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": "",
      },
    },
    "title": "{\\"oneProp\\":\\"yes!\\",\\"twoProp\\":\\"\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": "foobar",
      },
    },
    "title": "{\\"oneProp\\":\\"yes!\\",\\"twoProp\\":\\"foobar\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": "foobaz",
      },
    },
    "title": "{\\"oneProp\\":\\"yes!\\",\\"twoProp\\":\\"foobaz\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": "test",
      },
    },
    "title": "{\\"oneProp\\":\\"yes!\\",\\"twoProp\\":\\"test\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": "",
      },
    },
    "title": "{\\"oneProp\\":\\"no :(\\",\\"twoProp\\":\\"\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": "foobar",
      },
    },
    "title": "{\\"oneProp\\":\\"no :(\\",\\"twoProp\\":\\"foobar\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": "foobaz",
      },
    },
    "title": "{\\"oneProp\\":\\"no :(\\",\\"twoProp\\":\\"foobaz\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": "test",
      },
    },
    "title": "{\\"oneProp\\":\\"no :(\\",\\"twoProp\\":\\"test\\"}",
  },
]
`;

exports[`cartesian legends render title where props are nested 1`] = `
Array [
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": Object {
          "colors": Object {
            "bg": false,
            "fg": true,
          },
        },
      },
    },
    "title": "{\\"oneProp\\":\\"yes!\\",\\"twoProp\\":{\\"colors\\":{\\"fg\\":\\"yes!\\",\\"bg\\":\\"no :(\\"}}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": Object {
          "colors": Object {
            "bg": false,
            "fg": false,
          },
        },
      },
    },
    "title": "{\\"oneProp\\":\\"yes!\\",\\"twoProp\\":{\\"colors\\":{\\"fg\\":\\"no :(\\",\\"bg\\":\\"no :(\\"}}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": Object {
          "colors": Object {
            "bg": false,
            "fg": true,
          },
        },
      },
    },
    "title": "{\\"oneProp\\":\\"no :(\\",\\"twoProp\\":{\\"colors\\":{\\"fg\\":\\"yes!\\",\\"bg\\":\\"no :(\\"}}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": Object {
          "colors": Object {
            "bg": false,
            "fg": false,
          },
        },
      },
    },
    "title": "{\\"oneProp\\":\\"no :(\\",\\"twoProp\\":{\\"colors\\":{\\"fg\\":\\"no :(\\",\\"bg\\":\\"no :(\\"}}}",
  },
]
`;

exports[`cartesian legends render with null variants 1`] = `
Array [
  Object {
    "story": Object {
      "props": Object {
        "oneProp": 1,
        "twoProp": 2,
      },
    },
    "title": "{\\"oneProp\\":\\"one\\",\\"twoProp\\":\\"two\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": 1,
        "twoProp": null,
      },
    },
    "title": "{\\"oneProp\\":\\"one\\",\\"twoProp\\":\\"meh\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": null,
        "twoProp": 2,
      },
    },
    "title": "{\\"oneProp\\":\\"meh\\",\\"twoProp\\":\\"two\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": null,
        "twoProp": null,
      },
    },
    "title": "{\\"oneProp\\":\\"meh\\",\\"twoProp\\":\\"meh\\"}",
  },
]
`;

exports[`cartesian vanilla empty 1`] = `Array []`;

exports[`cartesian vanilla empty prop 1`] = `Array []`;

exports[`cartesian vanilla empty prop, complex 1`] = `Array []`;

exports[`cartesian vanilla multi prop 1`] = `
Array [
  Object {
    "story": Object {
      "props": Object {
        "oneProp": "one-1",
        "twoProp": "two",
      },
    },
    "title": "{\\"oneProp\\":\\"one-1\\",\\"twoProp\\":\\"two\\"}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": "one-2",
        "twoProp": "two",
      },
    },
    "title": "{\\"oneProp\\":\\"one-2\\",\\"twoProp\\":\\"two\\"}",
  },
]
`;

exports[`cartesian vanilla single prop 1`] = `
Array [
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
      },
    },
    "title": "{\\"oneProp\\":true}",
  },
]
`;

exports[`xproduct should product 1`] = `
Array [
  Array [
    1,
    3,
    4,
  ],
  Array [
    1,
    3,
    2,
  ],
  Array [
    1,
    2,
    4,
  ],
  Array [
    1,
    2,
    2,
  ],
  Array [
    2,
    3,
    4,
  ],
  Array [
    2,
    3,
    2,
  ],
  Array [
    2,
    2,
    4,
  ],
  Array [
    2,
    2,
    2,
  ],
]
`;

exports[`xproduct should product 2`] = `
Array [
  Array [
    1,
    3,
    4,
  ],
  Array [
    1,
    3,
    null,
  ],
  Array [
    1,
    null,
    4,
  ],
  Array [
    1,
    null,
    null,
  ],
  Array [
    null,
    3,
    4,
  ],
  Array [
    null,
    3,
    null,
  ],
  Array [
    null,
    null,
    4,
  ],
  Array [
    null,
    null,
    null,
  ],
]
`;
